File: src/zmq-client-demo.cpp
#include <chrono>
using namespace std::chrono_literals;
#include <thread>

#include <argparse/argparse.hpp>
#include <fmt/core.h>
#include <nlohmann/json.hpp>
// for convenience
using json = nlohmann::json;
using namespace nlohmann::literals; // for ""_json
#include <spdlog/spdlog.h>
#include <zmq.hpp>


auto main(int argc, char** argv) -> int {
	auto argv_parser = argparse::ArgumentParser(__FILE__, "0.1.0");
	argv_parser.add_argument("-p", "--port")
		.default_value(11111)
		.scan<'i', int>()
		.help(fmt::format("Port used for the zeromq PUB server, constraints: 0 < port <= {}",
						  std::pow(2, 16) - 1));

	try {
		argv_parser.parse_args(argc, argv);
	} catch (const std::exception& err) {
		fmt::println("{}", err.what());
		return 2;
	}

	const auto port = argv_parser.get<int>("port");

	auto	   zmq_ctx = zmq::context_t();
	auto	   subscriber = zmq::socket_t(zmq_ctx, zmq::socket_type::sub);
	const auto addr = fmt::format("tcp://*:{}", port);
	subscriber.connect(addr);
	const auto topic = std::string("/cars");
	subscriber.set(zmq::sockopt::subscribe, topic);
	spdlog::info("Created zeromq SUB socket connected to addr: {} listening on topic: {}", addr,
				 topic);

	while (true) {
		zmq::message_t message;
		subscriber.recv(&message, ZMQ_RCVMORE);
		std::string text(static_cast<char*>(message.data()), message.size());
		fmt::println("Received {}", text);
		std::this_thread::sleep_for(10ms);
	}

	return 0;
}

File: zmq-client-demo.py
#!/usr/bin/env -S pixi run python3

import os
import sys
import time
import argparse

import zmq
import cbor2
from loguru import logger
try:
    from rich import print, pretty
    pretty.install()
except ImportError or ModuleNotFoundError:
    pass


def main(argc: int, argv: list[str]) -> int:
    argv_parser = argparse.ArgumentParser(prog=os.path.basename(__file__).removesuffix(".py"),
                                         description="ZeroMQ client demo")
    argv_parser.add_argument("-p", "--port", type=int, default=5555, help="Port number")
    argv_parser.add_argument("-t", "--topic", type=str, default="streetlamps", help="Topic name", choices=["streetlamps", "cars"])

    args = argv_parser.parse_args(argv[1:])

    context = zmq.Context()

    logger.info(f"Connecting to server on port {args.port}...")
    subscriber = context.socket(zmq.SUB)
    subscriber.connect(f"tcp://localhost:{args.port}")
    # topic: bytes = b"cars"
    topic: bytes = args.topic.encode("utf-8")
    # topic: bytes = b"streetlamps"
    subscriber.setsockopt(zmq.SUBSCRIBE, topic)
    logger.info("Connected!")

    n_messages_received: int = 0
    try:
        while True:
            message = subscriber.recv()
            n_messages_received += 1
            data = cbor2.loads(message[len(topic):]) # skip the first n bytes as they are the topic name
            print(f"Received message #{n_messages_received}: {data}")
            time.sleep(0.1)
    except KeyboardInterrupt:
        logger.warning("Interrupted!")
    finally:
        subscriber.close()
        context.term()

    return 0


if __name__ == "__main__":
    sys.exit(main(len(sys.argv), sys.argv))
